Client hello spec from raw bytes

max-b opened the pull request:
This is a first draft pass at implementing the dynamic mimicry part of https://github.com/getlantern/lantern-internal/issues/3349

I'm a little nervous about the amount of changes and whether Sergey will want to incorporate it. He hasn't responded to https://github.com/refraction-networking/utls/issues/54 yet.

I'd also like to incorporate the mimicry implementation into an actual request a la the rest of the tests.

I wanted to get some early feedback from y'all so I thought I'd open this draft request.

Thanks y'all!

max-b commented:
Oh I think I'm also not using the correct convention for marking code as `UTLS` specific, so I want to make sure to revisit that.

forkner commented:
@max-b he mentioned an upcoming thesis defense in another ticket was keeping him from stuff, not sure that's done yet.

forkner commented on u_parrots.go:
```diff
@@ -938,3 +941,367 @@ func removeRC4Ciphers(s []uint16) []uint16 {
 }
+
+func FingerprintClientHello(data []byte) (*ClientHelloSpec, error) {
+	m := &clientHelloMsg{raw: data}
```
are you preserving the parsing of this because you're expecting we'll needs some pieces from this as well?

forkner replied:
oic, you use bits that are parsed in there to fill in the extensions and so forth.

max-b replied:
I think the other part is that it keeps the code as similar to the `unmarshal` code from `crypto/tls` as possible. If you remove the `// utls specific` sections you end up with the exact same code from https://github.com/refraction-networking/utls/blob/43c36d3c1f57546d5cbb05c066df7b5a78686c51/handshake_messages.go#L353

I'm not fully convinced that's worthwhile, but it seemed like it could be useful. fwiw that's the same reason I use the `parse()` closure (so that it can return a bool, same as the original code) :shrug:

max-b commented:
> @max-b he mentioned an upcoming thesis defense in another ticket was keeping him from stuff, not sure that's done yet.

Ah ok yeah that makes sense. He was pretty helpful and prompt with https://github.com/refraction-networking/utls/pull/51 so I'm feeling good about the chances that he'll weigh in eventually.

forkner commented:
@max-b Because of the padding issue you raised, I think it may also turn out not to be as generally useful upstream as we'd hope (since we cannot create something as general as a u_parrots entry... ).  I'm still hoping a domain-specific template is something do-able and useful to us, but if it's seeming like overkill or impossible, that's fair too.

forkner commented on u_parrots.go:
```diff
@@ -938,3 +941,367 @@ func removeRC4Ciphers(s []uint16) []uint16 {
+				m.earlyData = true
+
+				// utls specific
+				// TODO: read this into a generic extension
```
I'm not sure we would want to carry this one through unless we had a fresh hello for each request.  ie it wouldn't be a fixed value between any two requests even in the case that the early data it was communicating was (which would also be questionable).

forkner replied:
Since deleting these is not length preserving it may indicate to do this perfectly we would also need to force a new hello for each request.

max-b replied:
Yeah so one other option is that we return an error if we find that the hello has one of these fields that we can't really copy and we can't properly imitate.

max-b commented:
> @max-b Because of the padding issue you raised, I think it may also turn out not to be as generally useful upstream as we'd hope (since we cannot create something as general as a u_parrots entry... ). I'm still hoping a domain-specific template is something do-able and useful to us, but if it's seeming like overkill or impossible, that's fair too.

I'm still feeling optimistic - I think that some of these more esoteric fields will be rare in recorded hellos, especially if we're generating ourselves vs passive listening. We may just need to make sure we can handle failing to generate specs from recorded hellos (we'd have to decide what we want to do if that happens).

hwh33 commented on u_parrots_test.go:
```diff
@@ -0,0 +1,157 @@
+	checkUTLSFingerPrintClientHello(t, HelloIOS_12_1)
+	checkUTLSFingerPrintClientHello(t, HelloRandomized)
+	checkUTLSFingerPrintClientHello(t, HelloRandomizedALPN)
+	checkUTLSFingerPrintClientHello(t, HelloRandomizedNoALPN)
```
Nice these are some really awesome tests (commenting on your whole testing process).

max-b replied:
Thanks! Yeah I'd like to add some actual live checks against the openssl server (as it currently does with the other client hello specs), but I feel good about this being a minimal set.

max-b commented on u_parrots.go:
```diff
@@ -938,3 +941,367 @@ func removeRC4Ciphers(s []uint16) []uint16 {
+				// Ignore unknown extensions.
+
+				// utls specific
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &GenericExtension{extension, extData})
```
Yeah so at the very least, I think we'd want to carry through GREASE extensions (they're currently being handled by the `GenericExtension` approach).

It's a little hard to think abstractly about what kinds of other generic extensions might be and whether we'd want to copy/record them. Two thoughts on that front:
- Maybe we should try and do some larger data gathering on extensions that aren't parsed here?
- Maybe we should somewhat frequently send active queries and record client hellos so that we can update the spec as we see new hellos (so that changing hellos are accurately reflected??)

forkner replied:
I think a look through the extensions in the spec is worthwhile.  Also although we don't know the exact distribution or what might show up tomorrow, we can/should certainly test with the set of highly likely browsers especially in china ... chrome, uc browser, qq, 360 safe, sougou, safari, ie, edge, firefox, edge ... mobile variants.

forkner replied:
I'm not sure carrying through fixed grease values is the right idea if the hello is used as a template for multiple requests...  That's another sorta gray area where there appears to be /a/ behavior available (what BoringSSL does... random legal value for each new hello (with some caveats)) though it /could/ vary in others.

forkner replied:
It looks like your other stuff supports re-greasing, but I'm not sure this one does... I think assigning a new grease value depends specifically seeing a UTLSGreaseExtension in this case? https://github.com/getlantern/utls/blob/master/u_parrots.go#L627

hwh33 replied:
> Maybe we should try and do some larger data gathering on extensions that aren't parsed here?

> Maybe we should somewhat frequently send active queries and record client hellos so that we can update the spec as we see new hellos (so that changing hellos are accurately reflected??)

Definitely agree on both points.  I'll likely be testing lots of possible OS/browser combinations anyway, so that should give us a good sample to play with.  I think it would also make a lot of sense for the client to report any hellos it is unable to parse.  In that case, we could just have some reasonable defaults for the client to fallback to (these could be built-in, distributed by the config server, or both).  The client could send the hello along with browser and OS information.  We could collect these to study and figure out what we're missing.

I suspect that by testing a good range of browser and OS combinations ourselves pre-deployment, we will find most of the missing extensions.  The reports from the clients will probably help us clean up the majority of the rest, leaving us with a few rare edge cases.  Like you say, we also want to track any changes which lead to parsing errors and the reports from the clients will help us with that as well.

hwh33 replied:
And I think all of that together means that it's unlikely we'd ever want to blindly carry something through.

max-b replied:
> 

It looks like your other stuff supports re-greasing, but I'm not sure the extensions re-grease when using a GenericExtension... I think assigning a new grease value depends specifically seeing a UTLSGreaseExtension in this case? https://github.com/getlantern/utls/blob/master/u_parrots.go#L627

Good point - I can just check if the extension id is a grease value and make it a grease extension instead. Thanks for the note on that!

max-b replied:
> And I think all of that together means that it's unlikely we'd ever want to blindly carry something through.

I _thiiink_ there might be a couple other (non-GREASE, non-padding) places where I'm currently leaning on the `GenericExtension` in order to properly form the hello. Let me add some logging back in so I can double check that

max-b replied:
@forkner I've updated the code to properly handle the `UtlsGreaseExtension` instead of shoving it into a `GenericExtension` which I think will take care of regreasing for us. Thanks for that tip!

max-b commented on u_tls_extensions.go:
```diff
@@ -248,6 +248,7 @@ type ALPNExtension struct {
 func (e *ALPNExtension) writeToUConn(uc *UConn) error {
 	uc.config.NextProtos = e.AlpnProtocols
 	uc.HandshakeState.Hello.AlpnProtocols = e.AlpnProtocols
+	uc.HandshakeState.Hello.NextProtoNeg = len(e.AlpnProtocols) > 0
```
Yeah I'm not in love with how this has to happen, but go`s `clientHelloMsg` has a boolean `nextProtoNeg` field. [ALPN](https://tools.ietf.org/html/rfc7301) is an extension for negotiating an application layer protocol. The way that this is normally determined is by way of the underlying `UConn.config`:
https://github.com/getlantern/utls/blob/master/handshake_client.go#L77
```
		nextProtoNeg:                 len(config.NextProtos) > 0,
		alpnProtocols:                config.NextProtos,
```

However, we can't really modify the `config` when just generating the spec. In fact I find this whole logic flow/dependency really strange. I was struggling a bit to understand the interaction between ALPN and the `nextprotoneg` [extension](https://tools.ietf.org/html/draft-agl-tls-nextprotoneg-04). I think maybe this should either be removed or replaced with:
```
uc.HandshakeState.Hello.NextProtoNeg = uc.HandshakeState.Hello.NextProtoNeg || len(e.AlpnProtocols) > 0
```

If I remove it, it fails the tests that check against the `HelloRandomizedALPN` client ids. It's possible that's actually due to a misimplementation in the `generateRandomizedSpec` code. I can look into it a little more...

forkner replied:
I think this controls the next protocol negotiation extension, not alpn ?https://tools.ietf.org/html/draft-agl-tls-nextprotoneg-04

hwh33 replied:
That's interesting because it doesn't seem like you've made any changes which should affect the ALPN logic.  I guess I must be missing something.  Possibly something to flag and ask Sergey about when we try to get this merged upstream.

max-b commented on u_parrots_test.go:
```diff
@@ -0,0 +1,157 @@
+
+func assertEquality(t *testing.T, fieldName string, expected, actual interface{}) {
+	if kActual, ok := actual.(KeyShare); ok {
+		kExpected := expected.(KeyShare)
```
As implemented it's checking the exact same property field on the exact same struct, so I wasn't too worried about it, but yeah that makes sense.

hwh33 replied:
Okay, sorry I missed that.  Yeah not a big deal then

forkner commented on u_parrots.go:
```diff
@@ -938,3 +941,367 @@ func removeRC4Ciphers(s []uint16) []uint16 {
+		}
+
+		// utls specific
+		clientHelloSpec.CipherSuites = m.cipherSuites
```
The cipher suites may contain a GREASE value, which in many cases be expected to change between hellos.  In the parrots, the location of the grease value is replaced by GREASE_PLACEHOLDER, and a new GREASE value is selected when generating a new hello.  Same note as elsewhere, there is only one selection method available (the random selection of boringssl)... may be safe to assume.

forkner replied:
oic, you've adjusted the generation to select a new grease value when it's GREASE_PLACEHOLDER or any valid grease value?

max-b replied:
Yeah inside `ApplyPreset` (the function that takes the `ClientHelloSpec` and populates the hello struct), it's already doing the regreasing. I just had to change the check:
https://github.com/getlantern/utls/pull/6/files#diff-17a6357df708d056c23f2eb7f0c63069R615-R661

max-b commented:
Something that I'm kicking myself I didn't see earlier. On this page:
https://tlsfingerprint.io/id/cab2baf994ded9c9

You can click the `Click To Expand` button in the `uTLS generated code` section and it will actually output a `utls` `ClientHelloSpec`

So a couple interesting points that fall out of this:
1. I think that means that Sergey already has some parsing library already written which takes raw bytes and generates specs
2. We could probably just use the raw bytes->fingerprint->`ClientHelloSpec` pipeline that it seems he's already built
3. He's made some determinations about whether or not to include padding extensions and whether or not to add other unknown extensions (see the generated spec for https://tlsfingerprint.io/id/29099b6040851138)

I feel a _little_ silly that I put work into this considering that it may prove to be entirely a duplicate of work he's already done, but at the least I feel like I learned a bunch :shrug: :sweat_smile: 

For the moment, I think I'm going to pause work here and wait until we hear back from him. Does that sound about right to you @hwh33 and @forkner ?

hwh33 commented:
> You can click the Click To Expand button in the uTLS generated code section and it will actually output a utls ClientHelloSpec

Oh, good find.  I actually couldn't get the upload function to work - is it working for you?  I was trying to see if it accepted random hellos or just looks for fingerprints it recognizes.

If it does accept random hellos and you can get a `ClientHelloSpec` out of it, then yeah, I might pause for now.  If we don't hear from Sergey for a week or so though, it may make sense to finish this work up so we have something to use in the meantime.

hwh33 commented:
Just saw the discussion [here](https://github.com/refraction-networking/utls/issues/54#issuecomment-668067164).  It sounds like tlsfingerprint.io is just looking for fingerprints it recognizes.  I'm not sure that will work for us.  I suppose we could make it work by gathering hellos from clients and creating a database of fingerprint hashes mapped to `ClientHelloSpec`s.  The dynamic approach implemented here is more attractive to me though.

max-b commented:
> Just saw the discussion [here](https://github.com/refraction-networking/utls/issues/54#issuecomment-668067164). It sounds like tlsfingerprint.io is just looking for fingerprints it recognizes. I'm not sure that will work for us. I suppose we could make it work by gathering hellos from clients and creating a database of fingerprint hashes mapped to `ClientHelloSpec`s. The dynamic approach implemented here is more attractive to me though.

My read of it is that the tlsfingerprint.io site is actually dynamically creating `ClientHelloSpec`s from raw bytes. Maybe I misread that, but hopefully he'll send over the code sooner than later and we can see.

We _are_ pretty close to having what I think is a workable solution here, but there are still some corner cases that I don't exactly know how we'd handle. I do think that we can take some suggestions from tlsfingerprint.io 's `ClientHelloSpec` generation for some of these questions we've yet to answer.

One thing that @myleshorton mentioned is that our current approach re: client hellos is pretty broken. If I understand it correctly, per proxy, our clients either send totally random client hellos, or only send the exact same one (I may be misunderstanding that - if so definitely correct me here). That seems like a very large statistical flag which we could potentially address by implementing some of the more conservative takes we've outlined here (always using boringssl padding, not passing through any questionable extensions, etc). We might still produce technically identifiable `ClientHello`s, but they might be close enough to the real deal to buy us something.

hwh33 commented:
> We might still produce technically identifiable ClientHellos, but they might be close enough to the real deal to buy us something.

That's a great point.  Yeah I agree, any attempt at this, even if it ends up being a little naive, is probably an improvement.

max-b commented:
Ok so I attempted to update the PR to incorporate Sergey's code ideas and the feedback we've been going back and forth on.

I'm tempted to add an additional flag for something like `allowUnknownGenericExtensions`, where we just blindly add generic extensions for anything we don't recognize, but I think that could be an iterative process.

What do y'all think about the error approach? Basically it returns an error if any extensions exists that we don't think we can easily represent. I think that's probably the most conservative approach? @hwh33 mentioned some more sophisticated APIs, but I think this one might be good enough?

FWIW in case anyone hasn't seen it, here's the code that sergey shared with us:
https://drive.google.com/drive/folders/1nptA8aWlYVCONwwQEdXyX5Y04f_zkKef?usp=sharing

hwh33 commented:
> I'm tempted to add an additional flag for something like allowUnknownGenericExtensions, where we just blindly add generic extensions for anything we don't recognize, but I think that could be an iterative process.

That sounds good to me!

hwh33 commented on u_parrots.go:
```diff
@@ -938,3 +941,261 @@ func removeRC4Ciphers(s []uint16) []uint16 {
 	return s[:sliceLen]
 }
+
+// FingerprintClientHello returns a ClientHelloSpec which is based on the ClientHello that is passed in as the data argument
```
I think the documentation should also specify whether the record header should be present.  It looks like the answer is no.  Personally, I would say it makes the most sense to accept full records, but I think that's a subjective design decision.  Either way, it's important IMO to have that specified here.

max-b replied:
Really good point! (on both comments)

max-b replied:
I stuck with accepting the Handshake Message only, just because that's what `crypto/tls.clientHelloMsg.marshal()` outputs:
https://golang.org/src/crypto/tls/handshake_messages.go

I'm open to maybe additionally adding something like handling full tls records if you think that'd be useful?

I did go ahead and make the documentation more explicit.

hwh33 replied:
I think it is more likely that someone has a full record than just the handshake data.  OTOH, if someone had only the handshake data and we only accepted full records, they'd have to create their own header which is awkward.  Starting with the logic [here](https://github.com/getlantern/tlsutil/blob/7e08a8c42b7b063b83eecca03b84e1fa4086e9e4/hello.go#L22) may give you an idea of what's required to parse the header and read the data out of a record.  I'm sure you can simplify it quite a bit if you know that you're dealing with a client hello though.

hwh33 replied:
Yeah I think if you're confident that you've got a ClientHello record, you can just strip the first 5 bytes off.  So accepting just the record data is probably fine.

max-b replied:
@hwh33 thought you should see that Sergey agreed w/ you :grin: https://github.com/refraction-networking/utls/pull/58/files#r472576678

hwh33 replied:
Heh, thanks 😛

That's great to see that Sergey's already gone through the PR.  Looks like he's provided some good feedback as well.

max-b replied:
Yeah I'm amazed he was as thorough as he was as quickly as he was considering I think his thesis/dissertation defense is tomorrow???

hwh33 replied:
Yeah, no kidding!!

hwh33 commented on u_parrots.go:
```diff
@@ -938,3 +941,261 @@ func removeRC4Ciphers(s []uint16) []uint16 {
+			if isGREASEUint16(extension) {
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &UtlsGREASEExtension{unGREASEUint16(extension), extData})
+			} else {
+				return clientHelloSpec, fmt.Errorf("unsupported extension %x", extension)
```
Bit of a nit, but `%#x` formats like `0x10` rather than `10`, which would make it more clear that the number is hex.

myleshorton commented:
Have we run test hello bytes from a variety of browsers through this just to check how they do? I do think it would be a good idea to just run as many checks as possible to determine what extensions might commonly be out there that we don't recognize and/or can't just copy, as it might be straightforward to re-generate the common ones or add special handling for them.

hwh33 commented:
I'm definitely going to do that as part of the work in capturing hellos from different browsers.  I think we will also iterate on this to an extent by seeing what errors we get once we deploy this.  I actually just left some details on possible strategy [here](https://github.com/getlantern/lantern-internal/issues/3349#issuecomment-671627016).

max-b commented:
We've sort

> Have we run test hello bytes from a variety of browsers through this just to check how they do? I do think it would be a good idea to just run as many checks as possible to determine what extensions might commonly be out there that we don't recognize and/or can't just copy, as it might be straightforward to re-generate the common ones or add special handling for them.

This is also _sort of_ done for us in conjunction with Sergey and https://tlsfingerprint.io/. Of his list of [Top Fingerprints](https://tlsfingerprint.io/top/fingerprints), only 3 have a "generic"(unknown) extension that will currently fail our spec generation.
https://tlsfingerprint.io/id/5408690af1e08199
https://tlsfingerprint.io/id/247451152181eae6
https://tlsfingerprint.io/id/1fa228e941491e51

Those extensions are `0x0018`, `0x000f`, and `0x0029`.

Obviously data from the CU campus won't perfectly mirror our user's clients, but I think it's at least a good reassurance.

As for those other 3 unknown extensions, we can do one or more things:
- Research them and see if we can understand their usage and proper mimicry
- Pass them along as `GenericExtensions`. This _should_ be safe as long as our proxy servers tls implementations properly ignore them
- Continue to return errors in validation as a more conservative approach hoping that we'll see another `ClientHello` that _does_ validate later.

max-b commented:
> Nice, looking good @max-b. I left a couple of comments on minor things. Looks like you have a couple of lingering TODOs as well in case you weren't aware.

Thanks for the feedback!!

max-b commented:
I incorporated @hwh33's feedback and changed the error pattern. I didn't think it made sense to return a non-nil `ClientHelloSpec` when encountering an error because I don't think there's anything useful that can be done with that value.

hwh33 commented on u_parrots.go:
```diff
@@ -938,3 +941,263 @@ func removeRC4Ciphers(s []uint16) []uint16 {
+// FingerprintClientHello returns a ClientHelloSpec which is based on the ClientHello that is passed in as the data argument
+// If the ClientHello passed in has extensions that are not recognized or cannot be handled, it will return a non-nil error and a nil `*ClientHelloSpec` value
+// The `data` should be the ClientHello record as produced by `crypto/tls.clientHelloMsg.marshal()`
+// ie. it should not contain the full tls record, just the handshake message as outlined in https://tools.ietf.org/html/rfc5246#section-7.4
```
The backticks won't format in Godoc unfortunately.  Nor will the newlines as-is.  You need a blank line between the lines, which will create a new paragraph.  This is how the documentation currently renders:

![image](https://user-images.githubusercontent.com/4043597/89844480-ed0e2480-db38-11ea-83e5-b167291318c2.png)

[This is a kind of nifty primer on Godoc formatting](https://pkg.go.dev/github.com/natefinch/godocgo?tab=doc) if you're looking for more info.

max-b replied:
:shakes-fist-at-sky:!!!! Thanks for the tips!!!

max-b commented:
@hwh33 and @forkner I think I'm going to merge this here and then open it against Sergey's upstream unless y'all think I should hold off on one of those?

forkner commented on u_parrots.go:
```diff
@@ -938,3 +941,265 @@ func removeRC4Ciphers(s []uint16) []uint16 {
+					return nil, errors.New("SNI value may not include a trailing dot")
+				}
+
+				clientHelloSpec.Extensions = append(clientHelloSpec.Extensions, &SNIExtension{serverName})
```
If we're guessing on padding, is there still a reason to bake the specific name into the spec?

max-b replied:
Nope I don't think there is. Good point, I'll do that now.

max-b commented:
I also suspect that Sergey will have a lot of feedback on this if we open a PR against upstream (and he's a little slow w/ back and forth convos), so if that's the route we want to take, I think it might be better to do it sooner than later?

max-b commented:
I added some additional tests for checking extension equality and actually using a generated spec to create a connection with an `openssl` server

max-b commented:
I'm going to squash and merge this so that I have a clean commit I can use to open a PR against upstream. We already have some work to do for re-aligning w/ upstream `utls` and I think Sergey will have some feedback on this PR, so I think if there are ultimately further issues we want to work out/iterate on, it will be fine to do it from there.
